# STUDENT PERFORMANCE ANALYSIS

## Import Libraries


In [10]:
import pandas as pd




## Load Dataset


In [11]:
df = pd.read_csv("../data/StudentsPerformance.csv")

## Dataset Overview


In [12]:
df.shape

(1000, 8)

In [13]:
df.columns

Index(['gender', 'race/ethnicity', 'parental level of education', 'lunch',
       'test preparation course', 'math score', 'reading score',
       'writing score'],
      dtype='object')

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race/ethnicity               1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   math score                   1000 non-null   int64 
 6   reading score                1000 non-null   int64 
 7   writing score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


In [15]:
df.describe()

,math score,reading score,writing score
count,1000.00000,1000.000000,1000.000000
mean,66.08900,69.169000,68.054000
std,15.16308,14.600192,15.195657
min,0.00000,17.000000,10.000000
25%,57.00000,59.000000,57.750000
50%,66.00000,70.000000,69.000000
75%,77.00000,79.000000,79.000000
max,100.00000,100.000000,100.000000


## Data Cleaning

In [16]:
df.isnull().sum()

gender                         0
race/ethnicity                 0
parental level of education    0
lunch                          0
test preparation course        0
math score                     0
reading score                  0
writing score                  0
dtype: int64

In [20]:
df.duplicated().sum()

np.int64(0)

## Exploratory Data Analysis


### 5.1 Creating Average Score Feature


### Objective

The dataset contains separate scores for Mathematics, Reading, and Writing. To evaluate the overall academic performance of each student, a new feature called **average_score** is created by calculating the mean of these three subject scores.

In [32]:

df["average_score"] = (
    df["math score"] +
    df["reading score"] +
    df["writing score"]
) / 3

df.head()

,gender,race/ethnicity,parental level of education,lunch,test preparation course,math score,reading score,writing score,average_score
0,female,group B,bachelor's degree,standard,none,72,72,74,72.666667
1,female,group C,some college,standard,completed,69,90,88,82.333333
2,female,group B,master's degree,standard,none,90,95,93,92.666667
3,male,group A,associate's degree,free/reduced,none,47,57,44,49.333333
4,male,group C,some college,standard,none,76,78,75,76.333333


### Insight

A new feature named **average_score** was successfully created. This feature combines Mathematics, Reading, and Writing scores into a single metric, making it easier to compare overall student performance and perform further analysis.

### 5.2 Overall Student Performance


### Objective

This analysis aims to understand the overall academic performance of students by examining the average score, highest score, lowest score, and subject-wise average scores.

In [35]:
# Overall Average Score

overall_avg = df["average_score"].mean()
overall_avg

np.float64(67.77066666666666)

In [41]:
# Highest Average Score

highest_avg = df["average_score"].max()


In [40]:
# Lowest Average Score

lowest_avg = df["average_score"].min()


In [39]:
# Subject Wise Average Score

subject_avg = df[
    ["math score", "reading score", "writing score"]
].mean()


In [42]:
#OUTPUT
print(f"Overall Average Score: {overall_avg:.2f}")
print(f"Highest Average Score: {highest_avg:.2f}")
print(f"Lowest Average Score: {lowest_avg:.2f}")
print("\nSubject-wise Average Scores:")
print(subject_avg)

Overall Average Score: 67.77
Highest Average Score: 100.00
Lowest Average Score: 9.00

Subject-wise Average Scores:
math score       66.089
reading score    69.169
writing score    68.054
dtype: float64


### Insight

The overall average score of students is **67.77**, indicating a moderate level of academic performance across the dataset.

The highest average score achieved by a student is **100.00**, demonstrating exceptional performance across all three subjects, while the lowest average score is **9.00**, highlighting a significant variation in student achievement.

Among the three subjects, **Reading** has the highest average score (**69.17**), followed by **Writing** (**68.05**) and **Mathematics** (**66.09**). This suggests that students generally perform better in language-based subjects than in Mathematics.

### 5.3 Top Performing Students


### Objective

This analysis identifies the highest-performing students based on their overall average score. Examining top performers helps understand the level of academic excellence present in the dataset.

In [33]:


df.sort_values(
    by="average_score",
    ascending=False
).head(10)

,gender,race/ethnicity,parental level of education,lunch,test preparation course,math score,reading score,writing score,average_score
916,male,group E,bachelor's degree,standard,completed,100,100,100,100.000000
458,female,group E,bachelor's degree,standard,none,100,100,100,100.000000
962,female,group E,associate's degree,standard,none,100,100,100,100.000000
114,female,group E,bachelor's degree,standard,completed,99,100,100,99.666667
179,female,group D,some high school,standard,completed,97,100,100,99.000000
712,female,group D,some college,standard,none,98,100,99,99.000000
165,female,group C,bachelor's degree,standard,completed,96,100,100,98.666667
625,male,group D,some college,standard,completed,100,97,99,98.666667
903,female,group D,bachelor's degree,free/reduced,completed,93,100,100,97.666667
149,male,group E,associate's degree,free/reduced,completed,100,100,93,97.666667


### Insight

The analysis of the top-performing students reveals that several students achieved exceptionally high average scores across Mathematics, Reading, and Writing. The highest-performing student attained an average score of 100, demonstrating perfect performance in all subjects.

The presence of multiple high-scoring students indicates that while the overall average performance is moderate, a subset of students consistently excels across all academic areas.

### 5.4 Gender-wise Performance


### Objective

This analysis examines the academic performance of students based on gender. The goal is to identify any differences in Mathematics, Reading, Writing, and overall average scores.

In [43]:
gender_performance = df.groupby("gender")[
    ["math score", "reading score", "writing score", "average_score"]
].mean()

gender_performance

,math score,reading score,writing score,average_score
gender,,,,
female,63.633205,72.608108,72.467181,69.569498
male,68.728216,65.473029,63.311203,65.837483


### Insight

The gender-wise performance analysis reveals noticeable differences across subjects.

Female students achieved a higher overall average score (**69.57**) compared to male students (**65.84**). Females also outperformed males in both Reading (**72.61 vs 65.47**) and Writing (**72.47 vs 63.31**) scores.

However, male students scored higher in Mathematics, with an average score of **68.73** compared to **63.63** for female students.

These results suggest that female students generally perform better in language-related subjects, while male students show stronger performance in Mathematics. Despite the higher Mathematics scores among males, female students demonstrate superior overall academic performance due to significantly higher Reading and Writing scores.

### 5.5 Test Preparation Impact

### Objective

This analysis investigates the impact of completing a test preparation course on student academic performance. The goal is to determine whether participation in the course is associated with higher scores across subjects.

In [44]:
test_prep = df.groupby("test preparation course")[
    ["math score", "reading score", "writing score", "average_score"]
].mean()

test_prep.round(2)

,math score,reading score,writing score,average_score
test preparation course,,,,
completed,69.70,73.89,74.42,72.67
none,64.08,66.53,64.50,65.04


### Insight

The analysis shows a strong positive impact of test preparation on student performance.

Students who completed the test preparation course achieved an average score of **72.67**, whereas students who did not complete the course achieved an average score of **65.04**. This represents an improvement of approximately **7.63 points**.

Students who completed the course also scored higher across all subjects:
- Mathematics: 69.70 vs 64.08
- Reading: 73.89 vs 66.53
- Writing: 74.42 vs 64.50

These results suggest that participation in test preparation programs is associated with significantly better academic performance and may help students improve their understanding and examination skills.

### 5.6 Parental Education Impact

### Objective

This analysis examines the relationship between parental education level and student academic performance. The goal is to determine whether higher parental education is associated with better student outcomes.

In [45]:
parent_edu = df.groupby("parental level of education")[
    ["average_score"]
].mean().sort_values(
    by="average_score",
    ascending=False
)

parent_edu.round(2)

,average_score
parental level of education,
master's degree,73.60
bachelor's degree,71.92
associate's degree,69.57
some college,68.48
some high school,65.11
high school,63.10


### Insight

The analysis indicates a positive relationship between parental education level and student academic performance.

Students whose parents hold a **master's degree** achieved the highest average score (**73.60**), followed by those whose parents hold a **bachelor's degree** (**71.92**). In contrast, students whose parents have only a **high school education** recorded the lowest average score (**63.10**).

A gradual decline in student performance can be observed as parental education levels decrease. This trend suggests that higher parental education may contribute to a more supportive learning environment, greater academic guidance, and increased educational awareness.

Overall, the results indicate that parental education level is an important factor associated with student academic achievement.

### 5.7 Subject-wise Analysis

### Objective

This analysis compares student performance across Mathematics, Reading, and Writing to identify the strongest and weakest academic areas.

In [46]:
subject_avg = df[
    ["math score", "reading score", "writing score"]
].mean()

subject_avg.round(2)

math score       66.09
reading score    69.17
writing score    68.05
dtype: float64

### Insight

The subject-wise analysis shows that students performed best in Reading, with an average score of 69.17. Writing was the second-highest scoring subject with an average score of 68.05, while Mathematics had the lowest average score of 66.09.

The results indicate that students generally demonstrate stronger performance in language-related subjects compared to Mathematics. This may suggest that additional academic support and practice could be beneficial in improving mathematical performance.

### 5.8 Correlation Analysis

### Objective

This analysis examines the relationship between scores in Mathematics, Reading, and Writing to determine whether strong performance in one subject is associated with strong performance in others.

In [47]:
correlation_matrix = df[
    ["math score", "reading score", "writing score"]
].corr()

correlation_matrix

,math score,reading score,writing score
math score,1.000000,0.817580,0.802642
reading score,0.817580,1.000000,0.954598
writing score,0.802642,0.954598,1.000000


### Insight

The correlation analysis reveals a strong positive relationship between all three subjects.

Reading and Writing scores have the highest correlation (**0.95**), indicating that students who perform well in Reading tend to perform exceptionally well in Writing. This is expected because both subjects require similar language comprehension and communication skills.

Mathematics also shows a strong positive correlation with Reading (**0.82**) and Writing (**0.80**). These values suggest that students who perform well academically in one subject are generally likely to perform well in the others.

Overall, the results indicate that academic performance across subjects is closely related, with the strongest association observed between Reading and Writing.

## Visualizations


## Key Findings


## Conclusion
